File: typescript/list-ops/list-ops.ts
export default class List<T> {
  public values: T[];
  private _length: number;

  constructor(list?: T[]) {
    const newList = list || [];
    let length = 0;
    for (const _ of newList) {
      length++;
    }
    this.values = newList;
    this._length = length;
  }

  public length(): number {
    return this._length;
  }

  public push(value: T): List<T> {
    this._length++;
    this.values[this._length] = value;
    return this;
  }

  public unshift(value: T): List<T> {
    this.values = [value, ...this.values];
    this._length++;
    return this;
  }

  public append(list: List<T>): List<T> {
    for (const value of list.values) {
      this.values.push(value);
    }
    return this;
  }

  public concat(list: List<any>): List<any> {
    const listCopy = this;
    for (const item of list.values) {
      if (item.values) {
        listCopy.values.push(...item.values);
      } else {
        listCopy.values.push(item);
      }
    }
    return listCopy;
  }

  public filter(func: Function): List<{}> {
    const filteredList = new List();
    for (const value of this.values) {
      if (func(value) === true) {
        filteredList.values.push(value);
      }
    }
    return filteredList;
  }

  public map(func: Function): List<{}> {
    const mappedList = new List();
    for (const value of this.values) {
      mappedList.values.push(func(value));
    }
    return mappedList;
  }

  public reverse(): List<{}> {
    const reversed = new List();
    for (const value of this.values) {
      reversed.values.unshift(value);
    }
    return reversed;
  }

  public foldl(func: (acc: T, el: T) => T, initialVal: T): T {
    let acc = initialVal;
    for (const value of this.values) {
      acc = func(acc, value);
    }
    return acc;
  }

  public foldr(func: (acc: T, el: T) => T, initialVal: T): T {
    let acc = initialVal;
    for (let i = this.values.length - 1; i >= 0; i--) {
      acc = func(acc, this.values[i]);
    }
    return acc;
  }
}

File: go/list-ops/list_ops.go
package listops

// IntList is a collection of ints
type IntList []int

type (
	binFunc   func(int, int) int
	predFunc  func(int) bool
	unaryFunc func(int) int
)

// Foldr reduces a list from the right.
func (list IntList) Foldr(fn binFunc, accumulator int) int {
	for i := list.Length() - 1; i >= 0; i-- {
		accumulator = fn(list[i], accumulator)
	}
	return accumulator
}

// Foldl reduces a list from the left.
func (list IntList) Foldl(fn binFunc, accumulator int) int {
	for _, val := range list {
		accumulator = fn(accumulator, val)
	}
	return accumulator
}

// Filter returns a new list with items the predicate evaluated as true.
func (list IntList) Filter(fn predFunc) IntList {
	res := make(IntList, 0)
	for _, val := range list {
		if fn(val) {
			res = res.appendItem(val)
		}
	}
	return res
}

// Length returns the length of a list.
func (list IntList) Length() (length int) {
	for range list {
		length++
	}
	return length
}

// Map returns a new list with a function applied to each item.
func (list IntList) Map(fn unaryFunc) IntList {
	res := make(IntList, 0)
	for _, val := range list {
		res = res.appendItem(fn(val))
	}
	return res
}

// Reverse returns a new list with items in reverse order.
func (list IntList) Reverse() IntList {
	res := make(IntList, list.Length())
	idx := list.Length() - 1
	for i, val := range list {
		res[idx-i] = val
	}
	return res
}

// Append returns a new list with a list appended.
func (list IntList) Append(newList IntList) IntList {
	resLength := list.Length() + newList.Length()
	res := make(IntList, resLength)
	copy(res, list)
	idx := list.Length()
	for i, val := range newList {
		res[idx+i] = val
	}
	return res
}

func (list IntList) appendItem(item int) IntList {
	res := make(IntList, list.Length()+1)
	copy(res, list)
	res[list.Length()] = item
	return res
}

// Concat returns a new list of concatenated lists.
func (list IntList) Concat(newLists []IntList) IntList {
	res := make(IntList, list.Length())
	copy(res, list)
	for _, newList := range newLists {
		res = res.Append(newList)
	}
	return res
}
